### DISCUSSION TITLE
[.NET 9 + WASM + AOT] RELEASE no longer builds

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#18778] [General]
### Current behavior

So far I was building the RELEASE version of my solution using the 'use AOT as much as possible' mode `<WasmShellMonoRuntimeExecutionMode>InterpreterAndAOT</WasmShellMonoRuntimeExecutionMode>`

This no longer works but throws an error (screenshot below). 
Also: there are tons of warnings thrown additionally. How would I address these warnings?

You can reproduce by running `Minimal.Wasm/Make.bat` using this repo:
[Minimal.zip](https://github.com/user-attachments/files/17732604/Minimal.zip)

How could I create a RELEASE build with 'as much AOT as possible' and without warnings?

![image](https://github.com/user-attachments/assets/6ab03c19-bb5d-4004-891f-fe37cf6789e6)

**Comment by jeromelaban** [comment]
Thanks for the report. Make sure to follow the [directions from our release notes](https://platform.uno/docs/articles/migrating-from-previous-releases.html#net-9-rc2).

**Reply by jeromelaban** [reply]
The warnings are linker warnings that are not yet addressed, but are not a cause of concern. You can add those warnings to the ignore list for now.

**Reply by DierkDroth** [reply]
> Thanks for the report. Make sure to follow the [directions from our release notes](https://platform.uno/docs/articles/migrating-from-previous-releases.html#net-9-rc2).

@jeromelaban I failed to find references on the AOT mode with .NET9 in the (UNO) docs. Hence my original issue report. Could you please point me to relevant section(s) in the docs?

**Reply by jeromelaban** [reply]
See the net 9 [webassembly migration considerations](https://platform.uno/docs/articles/migrating-from-net8-to-net9.html#considerations-for-webassembly).

**Reply by DierkDroth** [reply]
> See the net 9 [webassembly migration considerations](https://platform.uno/docs/articles/migrating-from-net8-to-net9.html#considerations-for-webassembly).

Thanks @jeromelaban. Like I said: I have read these pages but could not find anything on how to build using the AOT mode. Where would I find the relevant info?

**Reply by jeromelaban** [reply]
All the information about net9 for webassembly is relevant to you, regardless of it being tagged as AOT or not. You need this specific section:

![image](https://github.com/user-attachments/assets/87e307da-990d-429e-8251-4920de5e7e2d)

**Reply by DierkDroth** [reply]
@jeromelaban I read through these pages and already have installed the 'WebAssembly tools from .NET' before I filed my original issue. I even read through [this](https://learn.microsoft.com/en-us/aspnet/core/release-notes/aspnetcore-9.0?view=aspnetcore-9.0). However, none of this tells me what I have to do with `<WasmShellMonoRuntimeExecutionMode>InterpreterAndAOT</WasmShellMonoRuntimeExecutionMode>` to get AOT as before (or better).

**Reply by jeromelaban** [reply]
Thanks, the information your provided was not enough to determine you ran this command.

In this case, this is an unknown .NET issue. Can you try running `dotnet workload list` ?

**Reply by DierkDroth** [reply]
Sure
![image](https://github.com/user-attachments/assets/d54e4110-4570-4abf-8cc8-c0a479486e1a)

**Reply by jeromelaban** [reply]
Then for some reason .NET does not find the proper wasm tools on your system, which is likely going to be an issue to open for the .NET team. Can you provide a binlog?

**Reply by DierkDroth** [reply]
binlog attached. 
Are you able to reproduce using my repo?

[msbuild.zip](https://github.com/user-attachments/files/17733784/msbuild.zip)

**Reply by jeromelaban** [reply]
Thanks. It looks like there's a behavior difference between RC2 and final bits when selecting the AOT compiler which needs fixing on uno's side.

Can you try adding:
```xml
<PropertyGroup>
   <RunAOTCompilation>true</RunAOTCompilation>
</PropertyGroup>
```

**Reply by jeromelaban** [reply]
Looking at it further, the reason this happens is because you're not using the Uno.Sdk. It's best if you migrate your projects to it to avoid those issues.

We'll update the documentation to mention that `<RunAOTCompilation>true</RunAOTCompilation>` is required for AOT when not using the Uno.Sdk.

**Reply by DierkDroth** [reply]
> Looking at it further, the reason this happens is because you're not using the Uno.Sdk. It's best if you migrate your projects to it to avoid those issues.
> 
> We'll update the documentation to mention that `<RunAOTCompilation>true</RunAOTCompilation>` is required for AOT when not using the Uno.Sdk.

Thanks for looking into @jeromelaban. 

To clarify: you're suggesting to replace `Sdk="Microsoft.NET.Sdk.WebAssembly`" in my projects by `Sdk="Uno.Sdk"`, correct?

Do I have to add a NuGet for this (which really would be something I wanted to avoid - to stay lean as possible on dependencies)?

Also: [here ](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/using-the-bootstrapper.html#upgrading-from-previous-versions-of-the-unowasmbootstrap-package) it reads that I have to have UNO SDK 5.5 or later. How would I do that?

**Reply by jeromelaban** [reply]
It's best to start from a blank template, using `dotnet new update` then `dotnet new unoapp -o MyBlankApp`, then migrate your code into it. The uno.sdk does not add release dependencies by default.

**Reply by DierkDroth** [reply]
@jeromelaban sorry, starting with a blank template is not an option, since it comes with a lot of stuff which I don't want (e.g. 'new' project structure) + would require massive restructuring on my code + many unknowns I'm not even aware off yet...

**Reply by DierkDroth** [reply]
@jeromelaban FYI quick update. This
```
<PropertyGroup>
   <RunAOTCompilation>true</RunAOTCompilation>
</PropertyGroup>
```
... got the repo above to build

**Reply by jeromelaban** [reply]
Thanks. This is the way forward, as at some point we'll stop supporting the non-uno.sdk setups. You can keep using it for now, but we strongly suggest to move to the single project structure.

**Reply by DierkDroth** [reply]
> Thanks. This is the way forward, as at some point we'll stop supporting the non-uno.sdk setups. You can keep using it for now, but we strongly suggest to move to the single project structure.

@jeromelaban so you know: this will give me a massive headache for no gain (I'm WASM only - and to some degree WinUI3 - which does what it should on all platforms by a single build - KISS)

**Reply by jeromelaban** [reply]
It's your choice. Single project is exactly about keeping projects simple, where the project does not have much moving parts to lose developers with.

**Reply by DierkDroth** [reply]
@jeromelaban well, the current structure of my project is plain and simple and - first and foremost - transparent. There is no stuff 'hidden' or 'done silently' nor is there any *.targets file or any other attempt to hide away complexity: a minimal set of files for clients (all platforms, landscape, portrait), server, multi-language support etc.
... whereas VS solution/project structure gets more and more obscure with every update ...

**Reply by DierkDroth** [reply]
@jeromelaban unfortunately my solution is still not building. I now have this section for RELEASE mode:
```
	<PropertyGroup Condition="'$(Configuration)'=='Release'">
		<DefineConstants>$(DefineConstants)</DefineConstants>
		<WasmPWAManifestFile>manifest.json</WasmPWAManifestFile>
		<WasmShellWebAppBasePath>/app/wasm/Latest/wwwroot/</WasmShellWebAppBasePath>
		<UnoXamlResourcesTrimming>true</UnoXamlResourcesTrimming>
		<RunAOTCompilation>true</RunAOTCompilation>
		<WasmShellILLinkerEnabled>true</WasmShellILLinkerEnabled>
	</PropertyGroup>
````
... and get the error in screenshot below.

 Here is the binlog:
[msbuild.zip](https://github.com/user-attachments/files/17737877/msbuild.zip)
I have disabled OneDrive or anything else which could access the directory (repo above builds no problem now).

Any idea what's going wrong?

![image](https://github.com/user-attachments/assets/28803c00-f96b-4b7e-a023-ca887d5ef9a2)

**Reply by jeromelaban** [reply]
This looks like a xaml trimming issue of some kind. Try disabling it with `UnoXamlResourcesTrimming`.

**Reply by jeromelaban** [reply]
>  LinkerConfig.xml(11,3): warning IL2007: Could not resolve assembly 'Microsoft.Extensions.Options'.

Yes, it can be removed, it's an assembly that has been removed long ago. The message is harmless.

**Reply by DierkDroth** [reply]
Perfect! Thanks @jeromelaban. I'll resume my testing and will report back...

**Reply by DierkDroth** [reply]
@jeromelaban update: **build succeeded** (with > 300 warnings, but still...) -> great

This leaves the question: what should do with `UnoXamlResourcesTrimming`? Is this no longer needed?

> Yes, it can be removed, it's an assembly that has been removed long ago. 

Also: not sure what you mean by this. I just checked, appears there even is a version for .NET9...

**Reply by DierkDroth** [reply]
> The warnings are linker warnings that are not yet addressed, but are not a cause of concern. You can add those warnings to the ignore list for now.

@jeromelaban I took another look at the gazillions of warnings when building and wanted to double check that they are not of concern. Many of those warning reference my code e.g. here
```
 C:\Users\dierk\source\repos\unusual-src\Unusual.Gui.Shared\Dashboard\ProjectsBySource.cs(164,5): Trim analysis warning IL2072: Unusual.Dashboard.ProjectsBySource.<.ctor>b__23_0(Object, EventArgs): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicParameterlessConstructor' in call to 'System.Activator.CreateInstance(Type)'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
 ```
... or here
```
C:\Users\dierk\source\repos\unusual-src\Unusual.Gui.Shared\Gui\PowerGrid.cs(2146,8): Trim analysis warning IL2057: Unusual.Gui.PowerGrid.NewRow(Object, Object, Boolean, Boolean): Unrecognized value passed to the parameter 'typeName' of method 'System.Type.GetType(String)'. It's not possible to guarantee the availability of the target type.
```
... and there are many more, even of different types

a) are you sure there is nothing to worry  about?
b) I'm actually surprised about these warning in the first place, since I have the section below in my LinkerConfig.xml - which are all of my assemblies. Are these linker exclusions no longer active?
```
<linker>
	<assembly fullname="Unusual.Common" />
	<assembly fullname="Unusual.Gui.Wasm" />
	<assembly fullname="Unusual.Wasm" />
...
```

**Reply by jeromelaban** [reply]
Trimming warnings in your code generally need to be fixed as they may reveal issues when the linker is enabled. The ones in Uno are generally false positives because Uno has been working with the linker enabled for the past five years, but we'll address them soon.

> This leaves the question: what should do with UnoXamlResourcesTrimming? Is this no longer needed?

It's needed, but it seems that there's an issue in your case but it's not clear why (we do not repro).

**Reply by DierkDroth** [reply]
> Trimming warnings in your code generally need to be fixed

That's what I though @jeromelaban. However, as per `LinkerConfig.xml` snippet above I thought the linker would just leave my 3 assemblies `Unusual.Common`, `Unusual.Gui.Wasm`, `Unusual.Wasm` similar to .NET 8 and UNO 5.1 - but apparently it doesn't. 

What am I doing  wrong?

> but it's not clear why (we do not repro).

Is there additional info I could provide?

**Reply by jeromelaban** [reply]
> That's what I though @jeromelaban. However, as per LinkerConfig.xml snippet above I thought the linker would just leave my 3 assemblies Unusual.Common, Unusual.Gui.Wasm, Unusual.Wasm similar to .NET 8 and UNO 5.1 - but apparently it doesn't.

It's not, but I don't have guidance to provide here, this is a linker behavior.

> Is there additional info I could provide?

A binlog would help.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban 

> It's not, but I don't have guidance to provide here, this is a linker behavior.

hmm ... so I understand:
- previously (.NET8) UNO was controlling the linker behavior by `LinkerConfig.xml`
- now (.NET 9) UNO no longer controls the linker behavior (since it falls in MS's court?!?) and it's not clear if `LinkerConfig.xml` has any relevance or not?
Am I understand this correctly?

> A binlog would help.

Does the binlog [here  ](https://github.com/unoplatform/uno/discussions/18778#discussioncomment-11244797)help?

**Reply by jeromelaban** [reply]
> now (.NET 9) UNO no longer controls the linker behavior (since it falls in MS's court?!?) and it's not clear if LinkerConfig.xml has any relevance or not?

The linker is controlled by LinkerConfig.xml, but the warnings are still raised. It's not a behavior we can change.

**Reply by DierkDroth** [reply]
> The linker is controlled by LinkerConfig.xml, but the warnings are still raised. It's not a behavior we can change.

I see. So my 3 assemblies (and others in `LinkerConfig.xml` as well) are in fact excluded - it's just that (useless and confusing) warnings would be thrown for even excluded assemblies, correct?

**Reply by DierkDroth** [reply]
> > The linker is controlled by LinkerConfig.xml, but the warnings are still raised. It's not a behavior we can change.
> 
> I see. So my 3 assemblies (and others in `LinkerConfig.xml` as well) are in fact excluded - it's just that (useless and confusing) warnings would be thrown for even excluded assemblies, correct?

For anyone interested: please find the [follow up here.](https://github.com/unoplatform/uno/discussions/18919)

**Reply by DierkDroth** [reply]
> > That's what I though @jeromelaban. However, as per LinkerConfig.xml snippet above I thought the linker would just leave my 3 assemblies Unusual.Common, Unusual.Gui.Wasm, Unusual.Wasm similar to .NET 8 and UNO 5.1 - but apparently it doesn't.
> 
> It's not, but I don't have guidance to provide here, this is a linker behavior.
> 
> > Is there additional info I could provide?
> 
> A binlog would help.

@jeromelaban does this binlog help?

[msbuild.zip](https://github.com/user-attachments/files/17949355/msbuild.zip)

**Reply by jeromelaban** [reply]
This binlog is still related to https://github.com/unoplatform/uno/discussions/18778#discussioncomment-11241481. You can fix it the same way.

**Reply by DierkDroth** [reply]
> This binlog is still related to [#18778 (reply in thread)](https://github.com/unoplatform/uno/discussions/18778#discussioncomment-11241481). You can fix it the same way.

@jeromelaban I'm not sure I understand.

To clarify: I have this section in my project file
```
	<PropertyGroup Condition="'$(Configuration)'=='Release'">
		<DefineConstants>$(DefineConstants)</DefineConstants>
		<NoWarn>$(NoWarn);IL2008;IL2026;IL2041;IL2057;IL2067;IL2070;IL2072;IL2075;IL2078;IL2087;IL2104</NoWarn>
		<RunAOTCompilation>true</RunAOTCompilation>
		<WasmPWAManifestFile>manifest.json</WasmPWAManifestFile>
		<WasmShellWebAppBasePath>/app/wasm/Latest/wwwroot/</WasmShellWebAppBasePath>
		<UnoXamlResourcesTrimming>true</UnoXamlResourcesTrimming>
		<WasmShellILLinkerEnabled>true</WasmShellILLinkerEnabled>
	</PropertyGroup>
```
... which throws these errors
![image](https://github.com/user-attachments/assets/6db024ee-ae74-48e3-9c44-5e6d3525f37f)

... and here is the BINLOG
[msbuild.zip](https://github.com/user-attachments/files/17961796/msbuild.zip)

What am I doing wrong?

**Reply by jeromelaban** [reply]
Thanks, the log you posted earlier did not have this information.

**Reply by jeromelaban** [reply]
Let's keep track here: https://github.com/unoplatform/uno/issues/18972. In the meantine, you can disable the Xaml Trimming.

**Reply by DierkDroth** [reply]
> Let's keep track here: #18972. In the meantine, you can disable the Xaml Trimming.

Thanks for taking care @jeromelaban

**Comment by DierkDroth** [comment]
Also: further down the error stack I see this error
` LinkerConfig.xml(11,3): warning IL2007: Could not resolve assembly 'Microsoft.Extensions.Options'.`

I suspect this is related to the respective line in the LinkerConfig.xml file (below). Could that line be removed safely? Or needed it to be replaced by something else?
```
<linker>
	<assembly fullname="Unusual.Common" />
	<assembly fullname="Unusual.Gui.Wasm" />
	<assembly fullname="Unusual.Wasm" />

	<assembly fullname="System.Core">
		<!-- This is required by JSon.NET and any expression.Compile caller -->
		<type fullname="System.Linq.Expressions*" />
	</assembly>

	<assembly fullname="Microsoft.Extensions.Options" />
	<assembly fullname="System.Runtime" />
	<assembly fullname="System.Threading" />
</linker>
```
